crates/runtime/src/init/dataset.rs

```diff
@@ -215,7 +215,15 @@ impl Runtime {
             Ok(data_connector) => {
                 // if connector marked as deferred, replace it with stub DeferredConnector
                 if data_connector.deferred_load() {
-                    Arc::new(DeferredConnector::new(data_connector)) as Arc<dyn DataConnector>
+                    // If the dataset has embeddings, datafusion needs to load the wrapped connector when it is ready
+                    if ds.has_embeddings() {
```

Comment:
A better interface would be to shift responsibility of wrapping the data connector to the dataset itself. So the dataset component would have a method like `ds.get_data_connector(data_connector) -> Arc<dyn DataConnector>` that would internally apply the EmbeddingConnector if required. That would also have prevented this bug in the first place - since it would have been more obvious that this logic needed to be applied.

As it is, this is turning into spaghetti code. :(

Reply:
Refactored to have a single method to create data connector and apply wrappers when needed. I'm not entierely sure if it would be beneficial to move it in the scope of `Dataset` type.